contracts/cw20-escrow/src/integration_test.rs

```diff
@@ -0,0 +1,141 @@
+
+use crate::msg::{CreateMsg, DetailsResponse, HandleMsg, InitMsg, QueryMsg, ReceiveMsg};
+
+fn mock_router() -> Router {
+    let env = mock_env();
+    let api = Box::new(MockApi::default());
+    let bank = SimpleBank {};
+
+    Router::new(api, env.block, bank, || Box::new(MockStorage::new()))
+}
+
+pub fn contract_escrow() -> Box<dyn Contract> {
+    let contract = ContractWrapper::new(
+        crate::contract::handle,
+        crate::contract::init,
+        crate::contract::query,
+    );
+    Box::new(contract)
+}
+
+pub fn contract_cw20() -> Box<dyn Contract> {
+    let contract = ContractWrapper::new(
+        cw20_base::contract::handle,
+        cw20_base::contract::init,
+        cw20_base::contract::query,
+    );
+    Box::new(contract)
+}
+
+#[test]
+// receive cw20 tokens and release upon approval
+fn escrow_happy_path_cw20_tokens() {
+    let mut router = mock_router();
+
+    // set personal balance
+    let owner = HumanAddr::from("owner");
+    let init_funds = coins(2000, "btc");
+    router
+        .set_bank_balance(owner.clone(), init_funds.clone())
+        .unwrap();
+
+    // set up cw20 contract with some tokens
+    let cw20_id = router.store_code(contract_cw20());
+    let msg = cw20_base::msg::InitMsg {
+        name: "Cash Money".to_string(),
+        symbol: "CASH".to_string(),
+        decimals: 2,
+        initial_balances: vec![Cw20CoinHuman {
+            address: owner.clone(),
+            amount: Uint128(5000),
+        }],
+        mint: None,
+    };
+    let cash_addr = router
+        .instantiate_contract(cw20_id, &owner, &msg, &[], "CASH")
+        .unwrap();
+
+    // set up reflect contract
+    let escrow_id = router.store_code(contract_escrow());
+    let escrow_addr = router
+        .instantiate_contract(escrow_id, &owner, &InitMsg {}, &[], "Escrow")
+        .unwrap();
+
+    // they are different
+    assert_ne!(cash_addr, escrow_addr);
+
+    // set up cw20 helpers
+    let cash = Cw20Contract(cash_addr.clone());
+
+    // ensure our balances
+    let owner_balance = cash.balance(&router, owner.clone()).unwrap();
+    assert_eq!(owner_balance, Uint128(5000));
+    let escrow_balance = cash.balance(&router, escrow_addr.clone()).unwrap();
+    assert_eq!(escrow_balance, Uint128(0));
+
+    // send some tokens to create an escrow
+    let arb = HumanAddr::from("arbiter");
+    let ben = HumanAddr::from("beneficiary");
+    let id = "demo".to_string();
+    let create_msg = ReceiveMsg::Create(CreateMsg {
+        id: id.clone(),
+        arbiter: arb.clone(),
+        recipient: ben.clone(),
+        end_height: None,
+        end_time: None,
+        cw20_whitelist: None,
+    });
+    let create_bin = to_binary(&create_msg).unwrap();
+    let send_msg = Cw20HandleMsg::Send {
+        contract: escrow_addr.clone(),
+        amount: Uint128(1200),
+        msg: Some(create_bin),
+    };
+    let res = router
+        .execute_contract(&cash_addr, &owner, &send_msg, &[])
```

Comment:
This is trivial, but maybe `.execute_contract(&owner, &cash_addr, ...)` is clearer, i. e. passing sender before recipient.

Reply:
Sounds good. Nice suggestion,